ws-client/src/transport.rs

```diff
@@ -184,74 +188,137 @@ impl<'a> WsTransportClientBuilder<'a> {
 			Mode::Plain => None,
 		};
 
+		self.try_connect(connector).await
+	}
+
+	async fn try_connect(
+		self,
+		mut tls_connector: Option<TlsConnector>,
+	) -> Result<(Sender, Receiver), WsHandshakeError> {
+		let mut target = self.target;
+		let mut used_sockaddrs = Vec::new();
+
 		let mut err = None;
-		for sockaddr in &self.target.sockaddrs {
-			match self.try_connect(*sockaddr, &connector).await {
-				Ok(res) => return Ok(res),
+
+		for _ in 0..MAX_REDIRECTIONS_ALLOWED {
+			log::debug!("Connecting to target: {:?}", target);
+
+			let sockaddr = match target.sockaddrs.pop() {
+				Some(addr) => {
+					used_sockaddrs.push(addr);
+					addr
+				}
+				None => return err.unwrap_or(Err(WsHandshakeError::NoAddressFound(target.host))),
+			};
+
+			let tcp_stream = match connect(sockaddr, self.timeout, &target.host, &tls_connector).await {
+				Ok(stream) => stream,
 				Err(e) => {
-					log::debug!("Failed to connect to sockaddr: {:?} with err: {:?}", sockaddr, e);
+					log::error!("Failed to connect to sockaddr: {:?}", sockaddr);
 					err = Some(Err(e));
+					continue;
 				}
+			};
+			let mut client = WsHandshakeClient::new(
+				BufReader::new(BufWriter::new(tcp_stream)),
+				&target.host_header,
+				&target.path_and_query,
+			);
+			if let Some(origin) = self.origin_header.as_ref() {
+				client.set_origin(origin);
```

Comment:
This breaks in `soketto` 0.7, not sure if upgrading in this PR is appropriate or not.